File: virtcontainers/types/capabilities.go
// Copyright (c) 2017 Intel Corporation
//
// SPDX-License-Identifier: Apache-2.0
//

package types

const (
	blockDeviceSupport = 1 << iota
	blockDeviceHotplugSupport
	multiQueueSupport
	fsSharingUnsupported
)

// Capabilities describe a virtcontainers hypervisor capabilities
// through a bit mask.
type Capabilities struct {
	flags uint
}

// IsBlockDeviceSupported tells if an hypervisor supports block devices.
func (caps *Capabilities) IsBlockDeviceSupported() bool {
	if caps.flags&blockDeviceSupport != 0 {
		return true
	}
	return false
}

// SetBlockDeviceSupport sets the block device support capability to true.
func (caps *Capabilities) SetBlockDeviceSupport() {
	caps.flags = caps.flags | blockDeviceSupport
}

// IsBlockDeviceHotplugSupported tells if an hypervisor supports hotplugging block devices.
func (caps *Capabilities) IsBlockDeviceHotplugSupported() bool {
	if caps.flags&blockDeviceHotplugSupport != 0 {
		return true
	}
	return false
}

// SetBlockDeviceHotplugSupport sets the block device hotplugging capability to true.
func (caps *Capabilities) SetBlockDeviceHotplugSupport() {
	caps.flags |= blockDeviceHotplugSupport
}

// IsMultiQueueSupported tells if an hypervisor supports device multi queue support.
func (caps *Capabilities) IsMultiQueueSupported() bool {
	if caps.flags&multiQueueSupport != 0 {
		return true
	}
	return false
}

// SetMultiQueueSupport sets the device multi queue capability to true.
func (caps *Capabilities) SetMultiQueueSupport() {
	caps.flags |= multiQueueSupport
}

// IsFsSharingSupported tells if an hypervisor supports host filesystem sharing.
func (caps *Capabilities) IsFsSharingSupported() bool {
	return caps.flags&fsSharingUnsupported == 0
}

// SetFsSharingUnsupported sets the host filesystem sharing capability to true.
func (caps *Capabilities) SetFsSharingUnsupported() {
	caps.flags |= fsSharingUnsupported
}

File: src/libs/kata-types/src/capabilities.rs
// Copyright (c) 2019-2022 Alibaba Cloud
// Copyright (c) 2019-2022 Ant Group
//
// SPDX-License-Identifier: Apache-2.0
//

use bitmask_enum::bitmask;

/// CapabilityBits
#[bitmask(u8)]
pub enum CapabilityBits {
    /// hypervisor supports use block device
    BlockDeviceSupport,
    /// hypervisor supports block device hotplug
    BlockDeviceHotplugSupport,
    /// hypervisor supports multi queue
    MultiQueueSupport,
    /// hypervisor supports filesystem share
    FsSharingSupport,
    /// hypervisor supports hybrid-vsock
    HybridVsockSupport,
    /// hypervisor supports memory hotplug probe interface
    GuestMemoryProbe,
}

/// Capabilities describe a virtcontainers hypervisor capabilities through a bit mask.
#[derive(Debug, Clone)]
pub struct Capabilities {
    /// Capability flags
    flags: CapabilityBits,
}

impl Default for Capabilities {
    fn default() -> Self {
        Self::new()
    }
}

impl Capabilities {
    /// new Capabilities struct
    pub fn new() -> Self {
        Capabilities {
            flags: CapabilityBits { bits: 0 },
        }
    }

    /// set CapabilityBits
    pub fn set(&mut self, flags: CapabilityBits) {
        self.flags = flags;
    }

    /// add CapabilityBits
    pub fn add(&mut self, flags: CapabilityBits) {
        self.flags |= flags;
    }

    /// is_block_device_supported tells if the hypervisor supports block devices.
    pub fn is_block_device_supported(&self) -> bool {
        self.flags.and(CapabilityBits::BlockDeviceSupport) != 0
    }

    /// is_block_device_hotplug_supported tells if the hypervisor supports block devices.
    pub fn is_block_device_hotplug_supported(&self) -> bool {
        self.flags.and(CapabilityBits::BlockDeviceHotplugSupport) != 0
    }

    /// is_multi_queue_supported tells if the hypervisor supports device multi queue support.
    pub fn is_multi_queue_supported(&self) -> bool {
        self.flags.and(CapabilityBits::MultiQueueSupport) != 0
    }

    /// is_hybrid_vsock_supported tells if an hypervisor supports hybrid-vsock support.
    pub fn is_hybrid_vsock_supported(&self) -> bool {
        self.flags.and(CapabilityBits::HybridVsockSupport) != 0
    }

    /// is_fs_sharing_supported tells if an hypervisor supports host filesystem sharing.
    pub fn is_fs_sharing_supported(&self) -> bool {
        self.flags.and(CapabilityBits::FsSharingSupport) != 0
    }

    /// is_mem_hotplug_probe_supported tells if the hypervisor supports hotplug probe interface
    pub fn is_mem_hotplug_probe_supported(&self) -> bool {
        self.flags.and(CapabilityBits::GuestMemoryProbe) != 0
    }
}

#[cfg(test)]
mod tests {
    use crate::capabilities::CapabilityBits;

    use super::Capabilities;

    #[test]
    fn test_set_hypervisor_capabilities() {
        let mut cap = Capabilities::new();
        assert!(!cap.is_block_device_supported());

        // test legacy vsock support
        assert!(!cap.is_hybrid_vsock_supported());

        // test set block device support
        cap.set(CapabilityBits::BlockDeviceSupport);
        assert!(cap.is_block_device_supported());
        assert!(!cap.is_block_device_hotplug_supported());

        // test set block device hotplug support
        cap.set(CapabilityBits::BlockDeviceSupport | CapabilityBits::BlockDeviceHotplugSupport);
        assert!(cap.is_block_device_hotplug_supported());
        assert!(!cap.is_multi_queue_supported());

        // test set multi queue support
        cap.set(
            CapabilityBits::BlockDeviceSupport
                | CapabilityBits::BlockDeviceHotplugSupport
                | CapabilityBits::MultiQueueSupport,
        );
        assert!(cap.is_multi_queue_supported());

        // test set host filesystem sharing support
        cap.set(
            CapabilityBits::BlockDeviceSupport
                | CapabilityBits::BlockDeviceHotplugSupport
                | CapabilityBits::MultiQueueSupport
                | CapabilityBits::FsSharingSupport,
        );
        assert!(cap.is_fs_sharing_supported());

        // test set hybrid-vsock support
        cap.add(CapabilityBits::HybridVsockSupport);
        assert!(cap.is_hybrid_vsock_supported());
        // test append capabilities
        cap.add(CapabilityBits::GuestMemoryProbe);
        assert!(cap.is_mem_hotplug_probe_supported());
        assert!(cap.is_fs_sharing_supported());
    }
}
